### DISCUSSION TITLE
Path not found when using fastify with many batched requests

### DISCUSSION THREAD

**Discussion by finnp** [trpc/trpc#3952] [Q&A] [answered]
### Provide environment information

```
  System:
    OS: macOS 13.1
    CPU: (10) arm64 Apple M1 Max
    Memory: 16.34 GB / 64.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.4.0 - ~/.nvm/versions/node/v18.4.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v18.4.0/bin/yarn
    npm: 8.12.1 - ~/.nvm/versions/node/v18.4.0/bin/npm
  Browsers:
    Chrome: 110.0.5481.177
    Safari: 16.2
  npmPackages:
    @tanstack/react-query: ^4.24.10 => 4.24.10
    @trpc/client: ^10.13.2 => 10.13.2
    @trpc/react-query: ^10.13.2 => 10.13.2
    @trpc/server: ^10.13.2 => 10.13.2
    react: 18.1.0 => 18.1.0
    typescript: ^4.9.4 => 4.9.5
```

### Describe the bug

Not sure if I want to track this as a client or a server bug, but I'm more worried about the server part.

I know this request down there is not very realistic and was a result of an error where the client repeatedly asked for the same data. However I don't think this should result in a `404` from fastify.

Here is the error that is happening:

{"level":30,"time":1678292845000,"pid":739816,"hostname":"ip-172-31-36-177","reqId":"req-4","msg":"Route GET:/trpc/outputMedia.get,outputMedia.getPending,outputMedia.get,outputMedia.getPending,outputMedia.get,outputMedia.getPending,outputMedia.get,outputMedia.getPending,outputMedia.get,outputMedia.getPending,outputMedia.get,outputMedia.getPending,outputMedia.get,outputMedia.getPending,outputMedia.get,outputMedia.getPending,outputMedia.get,outputMedia.getPending,outputMedia.get,outputMedia.getPending,outputMedia.get,outputMedia.getPending,outputMedia.get,outputMedia.getPending,outputMedia.get,outputMedia.getPending?batch=1&input=%7B%7D not found"}

Up until this point everything is fine, but this query then failed.

Here's my fastify server config
```ts
import initEnv from "./lib/env";
initEnv();

import { fastifyTRPCPlugin } from "@trpc/server/adapters/fastify";
import fastify from "fastify";
import { initFirebase } from "./lib/firebase";
import { mainRouter } from "./api";
import webHooks from "./http-routes/web-hooks";
import modelNotifications from "./http-routes/model-notifications";
import { createContext } from "./lib/trpc";
import {
  serializerCompiler,
  validatorCompiler,
} from "fastify-type-provider-zod";

initFirebase();

const server = fastify({ logger: true });

server.setValidatorCompiler(validatorCompiler);
server.setSerializerCompiler(serializerCompiler);

server.register(modelNotifications);
server.register(webHooks);

server.register(fastifyTRPCPlugin, {
  prefix: "/trpc",
  trpcOptions: { router: mainRouter, createContext },
});

server.get("/", () => "Running 🧘‍♀️");

server
  .listen({
    host: "0.0.0.0",
    port: 3000,
  })
  .catch((err) => {
    server.log.error(err);
    process.exit(1);
  });
```

### Link to reproduction

none

### To reproduce

Not sure how to reproduce, but happens from time to time. Maybe someone has an idea just given the request.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

**Comment by Nick-Lucas** [comment]
Have you enabled batching on your client but not your server by any chance?

Can you share your setup on each side?

**Comment by finnp** [comment]
Yes I'm using the `httpBatchLink`

```ts
import {MainRouter} from '@api/index';
import {API_URL} from '@env';
import {firebase} from '@react-native-firebase/auth';
import {QueryClient, QueryClientProvider} from '@tanstack/react-query';
import {httpBatchLink, loggerLink} from '@trpc/client';
import {createTRPCReact} from '@trpc/react-query';

export const trpc = createTRPCReact<MainRouter>();

const trpcClient = trpc.createClient({
  links: [
    loggerLink({
      enabled: () => true,
    }),
    httpBatchLink({
      url: `${API_URL}/trpc`,
      async headers() {
        const token = await firebase.auth().currentUser?.getIdToken();
        return {
          Authorization: `Bearer ${token}`,
        };
      },
    }),
  ],
});

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      staleTime: 7 * 24 * 60 * 60 * 1000,
    },
    mutations: {
      onSettled(data, error) {
        if (error) return;

        if (
          data &&
          typeof data === 'object' &&
          'invalidates' in data &&
          Array.isArray(data.invalidates)
        ) {
          data.invalidates.forEach(queryKey => {
            console.log('Invalidate:', queryKey);
            queryClient.invalidateQueries({
              queryKey: [queryKey.split('.')],
            });
          });
        }
      },
    },
  },
});

export function TrpcProvider({children}: {children: React.ReactNode}) {
  return (
    <trpc.Provider client={trpcClient} queryClient={queryClient}>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </trpc.Provider>
  );
}
```

**Reply by Nick-Lucas** [reply]
So batching needs to be enabled on your backend too, or switch to httpLink of the client, but they can't be mixed

**Reply by finnp** [reply]
I think it's enabled by default. But I just found the likely ssue:

According to fastify 100 is the default max length for request:

> ## maxParamLength
> 
> You can set a custom length for parameters in parametric (standard, regex, and multi) routes by using maxParamLength option; the default value is 100 characters.
> 
> This can be useful especially if you have a regex-based route, protecting you against [DoS attacks](https://www.owasp.org/index.php/Regular_expression_Denial_of_Service_-_ReDoS).
> 
> **If the maximum length limit is reached, the not found route will be invoked.**
> 
> [¶](https://www.fastify.io/docs/latest/Reference/Server/#bodylimit)

But that requests is larger than this setting.

**Reply by finnp** [reply]
It's even mentioned in the TRPC docs: https://trpc.io/docs/fastify#create-fastify-server. I just didn't see it.

**Reply by Nick-Lucas** [reply]
You are right, it looks to be on by default, and makes sense what you've found. 100 really isn't much is it? :D

**Reply by Nick-Lucas** [reply]
@finnp do you know is that error originating within Fastify itself before even reaching tRPC, or it's getting truncated and tRPC returns 404 then fastify maps that to this error?

If it's the latter I do think there's a ticket in here to put something clearer in your console, but obviously fastify itself is out of our control :)

**Reply by Nick-Lucas** [reply]
Ah from fastify docs

> If the maximum length limit is reached, the not found route will be invoked.

Probably answers that

**Reply by finnp** [reply]
Yap exactly! I didn't exactly verify it but it just sounds exactly like what is supposed to happen from fastify's side. They wouldn't even give the request to trpc.

**Comment by finnp** [comment, marked as answer]
For clarity commenting here again to mark as answer.

Fastify by default only allows parameters in Urls to be 100 characters (see https://www.fastify.io/docs/latest/Reference/Server/#bodylimit)

So I need to increase this limit as mentioned here: https://trpc.io/docs/fastify#create-fastify-server
